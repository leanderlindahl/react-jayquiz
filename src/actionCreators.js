import fetch from 'cross-fetch';
import {
  INVALIDATE_CATEGORY,
  RECEIVE_QUESTIONS,
  REQUEST_QUESTIONS,
  SELECT_CATEGORY,
  // SET_CURRENT_ANSWER,
  SET_CURRENT_ANSWER_STATUS,
  SET_CURRENT_QUESTION_INDEX,
  SET_DISPLAY_ANSWER_RESPONSE,
  SET_GAME_OVER,
  SET_OPTIONS_DISABLED,
  SET_OUT_OF_TIME,
  SET_SCORE,
  SET_SELECTED_OPTION
} from './actions';

export function invalidateCategory(category) {
  return { type: INVALIDATE_CATEGORY, category };
}
export function selectCategory(category) {
  return { type: SELECT_CATEGORY, category };
}
export function requestQuestions(category) {
  return { type: REQUEST_QUESTIONS, category };
}
export function receiveQuestions(category, json) {
  return {
    type: RECEIVE_QUESTIONS,
    category,
    questions: json.results,
    receivedAt: Date.now()
  };
}
export function fetchQuestions(category) {
  return dispatch => {
    dispatch(requestQuestions(category));
    return fetch(`https://opentdb.com/api.php?amount=5&category=${category}&difficulty=easy`)
      .then(
        response => response.json(),
        error => console.error('An error occured when fetching the questions.') // eslint-disable-line no-console
      )
      .then(json => dispatch(receiveQuestions(category, json)));
  };
}
// export function setCurrentAnswer(currentAnswer) {
//   return { type: SET_CURRENT_ANSWER, payload: currentAnswer };
// }
export function setCurrentAnswerStatus(value) {
  return { type: SET_CURRENT_ANSWER_STATUS, payload: value };
}
export function setCurrentQuestionIndex(index) {
  return { type: SET_CURRENT_QUESTION_INDEX, payload: index };
}
export function setDisplayAnswerResponse(value) {
  return { type: SET_DISPLAY_ANSWER_RESPONSE, payload: value };
}
export function setGameOver(value) {
  return { type: SET_GAME_OVER, payload: value };
}
export function setOptionsDisabled(state) {
  return { type: SET_OPTIONS_DISABLED, payload: state };
}
export function setOutOfTime(state) {
  return { type: SET_OUT_OF_TIME, payload: state };
}
export function setScore(score) {
  return { type: SET_SCORE, payload: score };
}
export function setSelectedOption(selected) {
  return { type: SET_SELECTED_OPTION, payload: selected };
}
