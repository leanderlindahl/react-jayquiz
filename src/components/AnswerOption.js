import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Radio, Input } from 'antd';
import unescapeHTML from '../helpers/unescapeHTML';

const RadioGroup = Radio.Group;

class AnswerOption extends Component {
  handleRadioChange = () => {
    this.setState({ checked: true });
  };

  render() {
    const { option, onAnswerSelected, currentAnswerStatus, currentAnswer } = this.props;
    return (
      <li className={`answerOption ${currentAnswerStatus !== '' ? 'xgray' : null}`}>
        <Radio
          id={option}
          value={unescapeHTML(option)}
          onClick={onAnswerSelected}
          disabled={!currentAnswerStatus === ''}
          checked={currentAnswer === option}
          onChange={this.handleRadioChange}
        />
        <label htmlFor={option} className="radio-custom-label">
          {unescapeHTML(option)}
        </label>
      </li>
    );
  }
}

AnswerOption.propTypes = {
  currentAnswer: PropTypes.string,
  currentAnswerStatus: PropTypes.string.isRequired,
  onAnswerSelected: PropTypes.func.isRequired,
  option: PropTypes.string.isRequired
};

AnswerOption.defaultProps = {
  currentAnswer: ''
};

const mapStateToProps = state => ({
  currentAnswer: state.currentAnswer,
  currentAnswerStatus: state.currentAnswerStatus
});

export default connect(mapStateToProps)(AnswerOption);
