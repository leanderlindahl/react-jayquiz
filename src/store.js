import { createStore, compose, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import reducer from './reducers';

// const store = createStore(
//   reducer,
//   compose(
//     applyMiddleware(thunk),
//     window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
//   )
// );

/* eslint-disable no-underscore-dangle */
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
/* eslint-enable */

const store = createStore(reducer, composeEnhancers(applyMiddleware(thunk)));

// const store = createStore(reducer, compose(applyMiddleware(thunk)));
/* eslint-enable */

export default store;
