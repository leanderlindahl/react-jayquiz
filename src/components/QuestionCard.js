import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import {
  setCurrentAnswer,
  setCurrentAnswerStatus,
  setCurrentQuestionIndex,
  setDisplayAnswerResponse,
  setGameOver,
  setScore
} from '../actionCreators';
import AnswerOption from './AnswerOption';
import unescapeHTML from '../helpers/unescapeHTML';
import shuffleArray from '../helpers/shuffleArray';

class QuestionCard extends Component {
  constructor(props) {
    super(props);

    this.state = {
      options: []
    };
  }

  componentDidMount() {
    this.prepareOptions();
  }

  componentDidUpdate(prevProps) {
    const { question } = this.props;
    if (question !== prevProps.question) {
      this.prepareOptions();
    }
  }

  handleSelectedAnswer = event => {
    const { handleAnswerState, score, question } = this.props;
    const answer = unescapeHTML(question.correct_answer);
    const selected = unescapeHTML(event.target.value);

    if (selected === answer) {
      handleAnswerState(score + 1, 'right', selected);
    } else {
      handleAnswerState(score, 'wrong', selected);
    }
  };

  handleClickNext = () => {
    const { currentQuestionIndex, handleShowNext } = this.props;
    handleShowNext(currentQuestionIndex);
  };

  prepareOptions() {
    const { question } = this.props;
    const options = shuffleArray(question.incorrect_answers.concat(question.correct_answer));
    this.setState({
      options
    });
  }

  render() {
    const {
      currentAnswerStatus,
      displayAnswerResponse,
      handleShowResult,
      next,
      score,
      question,
      questionNumber
    } = this.props;

    const { options } = this.state;

    return (
      <div className="question-wrapper">
        <h3 className="score">
          Question
          {` ${questionNumber} - `}
          Score:
          {` ${score}`}
        </h3>
        <div className="question">{unescapeHTML(question.question)}</div>
        <div className="answers">
          <ul className="options">
            {options.map(option => (
              <AnswerOption
                key={option}
                option={option}
                onAnswerSelected={this.handleSelectedAnswer}
              />
            ))}
          </ul>
        </div>
        ======================================================
        <div className="actions">
          {displayAnswerResponse ? (
            <>
              <div>
                The answer was:
                {` ${currentAnswerStatus}
                The correct answer is ${unescapeHTML(question.correct_answer)}`}
              </div>
              {next ? (
                <button type="button" onClick={this.handleClickNext}>
                  Next question
                </button>
              ) : (
                <button type="button" onClick={handleShowResult}>
                  Result
                </button>
              )}
            </>
          ) : null}
        </div>
      </div>
    );
  }
}

QuestionCard.propTypes = {
  currentAnswerStatus: PropTypes.string,
  currentQuestionIndex: PropTypes.number,
  displayAnswerResponse: PropTypes.bool,
  handleAnswerState: PropTypes.func.isRequired,
  handleShowNext: PropTypes.func.isRequired,
  handleShowResult: PropTypes.func.isRequired,
  next: PropTypes.bool,
  score: PropTypes.number,
  question: PropTypes.object.isRequired,
  questionNumber: PropTypes.number
};

QuestionCard.defaultProps = {
  currentAnswerStatus: '',
  currentQuestionIndex: 0,
  displayAnswerResponse: false,
  next: false,
  score: 0,
  questionNumber: 1
};

const mapStateToProps = state => ({
  currentAnswerStatus: state.currentAnswerStatus,
  currentQuestionIndex: state.currentQuestionIndex,
  displayAnswerResponse: state.displayAnswerResponse,
  score: state.score
});

const mapDispatchToProps = dispatch => ({
  handleAnswerState(score, answerStatus, selected) {
    dispatch(setScore(score));
    dispatch(setCurrentAnswerStatus(answerStatus));
    dispatch(setCurrentAnswer(selected));
    dispatch(setDisplayAnswerResponse(true));
  },
  handleShowNext(index) {
    dispatch(setCurrentAnswerStatus(''));
    dispatch(setDisplayAnswerResponse(false));
    dispatch(setCurrentQuestionIndex(index + 1));
  },
  handleShowResult() {
    dispatch(setGameOver(true));
  }
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(QuestionCard);
