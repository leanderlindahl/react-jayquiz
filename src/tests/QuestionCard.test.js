import React from 'react';
import { shallow, render, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Provider } from 'react-redux';
import store from '../store';
// import setUpTests from './setUpTests';
import QuestionCard, { Unwrapped as UnwrappedQuestionCard } from '../components/QuestionCard';
import AnswerOption from '../components/AnswerOption';
import { setDisplayAnswerResponse } from '../actionCreators';

configure({ adapter: new Adapter() });

const currentAnswer = '';
const currentAnswerStatus = '';
const displayAswerResponse = false;
const formattedAnswer = '8';
const formattedQuestion = 'What amount of bits commonly equals one byte?';
const next = true;
const questionNumber = 1;
const score = 0;

const mockHandleAnswerState = jest.fn();
const mockHandleShowNext = jest.fn();
const mockHandleShowResult = jest.fn();
const mockOptions = ['1', '2', '64', '8'];
const mockQuestion = {
  category: 'Science: Computers',
  correct_answer: '8',
  difficulty: 'easy',
  incorrect_answers: ['1', '2', '64'],
  question: 'What amount of bits commonly equals one byte?',
  type: 'multiple'
};

// This test doesn't work when options are shuffled in the component
xtest('QuestionCard renders correctly', () => {
  const component = shallow(
    <UnwrappedQuestionCard
      handleAnswerState={mockHandleAnswerState}
      handleShowNext={mockHandleShowNext}
      handleShowResult={mockHandleShowResult}
      question={mockQuestion}
    />
  );
  expect(component).toMatchSnapshot();
});

test('AnswerOptions should equal given amount of options', () => {
  const component = shallow(
    <UnwrappedQuestionCard
      handleAnswerState={mockHandleAnswerState}
      handleShowNext={mockHandleShowNext}
      handleShowResult={mockHandleShowResult}
      question={mockQuestion}
    />
  );
  expect(component.find(AnswerOption).length).toEqual(mockOptions.length);
});

test('Answer response should be visible after question was answered', () => {
  store.dispatch(setDisplayAnswerResponse(true));
  const component = render(
    <Provider store={store}>
      <QuestionCard
        handleAnswerState={mockHandleAnswerState}
        handleShowNext={mockHandleShowNext}
        handleShowResult={mockHandleShowResult}
        question={mockQuestion}
      />
    </Provider>
  );
  expect(component.find('.answer-response').length).toEqual(1);
});
