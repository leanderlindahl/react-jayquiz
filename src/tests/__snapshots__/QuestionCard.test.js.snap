// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionCard renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <QuestionCard
    currentAnswerStatus=""
    currentQuestionIndex={0}
    displayAnswerResponse={false}
    handleAnswerState={[MockFunction]}
    handleShowNext={[MockFunction]}
    handleShowResult={[MockFunction]}
    next={false}
    question={
      Object {
        "category": "Science: Computers",
        "correct_answer": "8",
        "difficulty": "easy",
        "incorrect_answers": Array [
          "1",
          "2",
          "64",
        ],
        "question": "What amount of bits commonly equals one byte?",
        "type": "multiple",
      }
    }
    questionNumber={1}
    score={0}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3
          className="score"
        >
          Question
           1 - 
          Score:
           0
        </h3>,
        <div
          className="question"
        >
          What amount of bits commonly equals one byte?
        </div>,
        <div
          className="answers"
        >
          <ul
            className="options"
          >
            <Connect(AnswerOption)
              onAnswerSelected={[Function]}
              option="64"
            />
            <Connect(AnswerOption)
              onAnswerSelected={[Function]}
              option="2"
            />
            <Connect(AnswerOption)
              onAnswerSelected={[Function]}
              option="8"
            />
            <Connect(AnswerOption)
              onAnswerSelected={[Function]}
              option="1"
            />
          </ul>
        </div>,
        "======================================================",
        <div
          className="actions"
        />,
      ],
      "className": "question-wrapper",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Question",
            " 1 - ",
            "Score:",
            " 0",
          ],
          "className": "score",
        },
        "ref": null,
        "rendered": Array [
          "Question",
          " 1 - ",
          "Score:",
          " 0",
        ],
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "What amount of bits commonly equals one byte?",
          "className": "question",
        },
        "ref": null,
        "rendered": "What amount of bits commonly equals one byte?",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <ul
            className="options"
          >
            <Connect(AnswerOption)
              onAnswerSelected={[Function]}
              option="64"
            />
            <Connect(AnswerOption)
              onAnswerSelected={[Function]}
              option="2"
            />
            <Connect(AnswerOption)
              onAnswerSelected={[Function]}
              option="8"
            />
            <Connect(AnswerOption)
              onAnswerSelected={[Function]}
              option="1"
            />
          </ul>,
          "className": "answers",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Connect(AnswerOption)
                onAnswerSelected={[Function]}
                option="64"
              />,
              <Connect(AnswerOption)
                onAnswerSelected={[Function]}
                option="2"
              />,
              <Connect(AnswerOption)
                onAnswerSelected={[Function]}
                option="8"
              />,
              <Connect(AnswerOption)
                onAnswerSelected={[Function]}
                option="1"
              />,
            ],
            "className": "options",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "64",
              "nodeType": "class",
              "props": Object {
                "onAnswerSelected": [Function],
                "option": "64",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "2",
              "nodeType": "class",
              "props": Object {
                "onAnswerSelected": [Function],
                "option": "2",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "8",
              "nodeType": "class",
              "props": Object {
                "onAnswerSelected": [Function],
                "option": "8",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "class",
              "props": Object {
                "onAnswerSelected": [Function],
                "option": "1",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "ul",
        },
        "type": "div",
      },
      "======================================================",
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": null,
          "className": "actions",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3
            className="score"
          >
            Question
             1 - 
            Score:
             0
          </h3>,
          <div
            className="question"
          >
            What amount of bits commonly equals one byte?
          </div>,
          <div
            className="answers"
          >
            <ul
              className="options"
            >
              <Connect(AnswerOption)
                onAnswerSelected={[Function]}
                option="64"
              />
              <Connect(AnswerOption)
                onAnswerSelected={[Function]}
                option="2"
              />
              <Connect(AnswerOption)
                onAnswerSelected={[Function]}
                option="8"
              />
              <Connect(AnswerOption)
                onAnswerSelected={[Function]}
                option="1"
              />
            </ul>
          </div>,
          "======================================================",
          <div
            className="actions"
          />,
        ],
        "className": "question-wrapper",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Question",
              " 1 - ",
              "Score:",
              " 0",
            ],
            "className": "score",
          },
          "ref": null,
          "rendered": Array [
            "Question",
            " 1 - ",
            "Score:",
            " 0",
          ],
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "What amount of bits commonly equals one byte?",
            "className": "question",
          },
          "ref": null,
          "rendered": "What amount of bits commonly equals one byte?",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <ul
              className="options"
            >
              <Connect(AnswerOption)
                onAnswerSelected={[Function]}
                option="64"
              />
              <Connect(AnswerOption)
                onAnswerSelected={[Function]}
                option="2"
              />
              <Connect(AnswerOption)
                onAnswerSelected={[Function]}
                option="8"
              />
              <Connect(AnswerOption)
                onAnswerSelected={[Function]}
                option="1"
              />
            </ul>,
            "className": "answers",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Connect(AnswerOption)
                  onAnswerSelected={[Function]}
                  option="64"
                />,
                <Connect(AnswerOption)
                  onAnswerSelected={[Function]}
                  option="2"
                />,
                <Connect(AnswerOption)
                  onAnswerSelected={[Function]}
                  option="8"
                />,
                <Connect(AnswerOption)
                  onAnswerSelected={[Function]}
                  option="1"
                />,
              ],
              "className": "options",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "64",
                "nodeType": "class",
                "props": Object {
                  "onAnswerSelected": [Function],
                  "option": "64",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "2",
                "nodeType": "class",
                "props": Object {
                  "onAnswerSelected": [Function],
                  "option": "2",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "8",
                "nodeType": "class",
                "props": Object {
                  "onAnswerSelected": [Function],
                  "option": "8",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "1",
                "nodeType": "class",
                "props": Object {
                  "onAnswerSelected": [Function],
                  "option": "1",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "ul",
          },
          "type": "div",
        },
        "======================================================",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": null,
            "className": "actions",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
